tests/async_tests/test_interfaces.py

```diff
@@ -182,14 +183,37 @@
 async def test_http_proxy(
     proxy_server: typing.Tuple[bytes, bytes, int], proxy_mode: str
 ) -> None:
-    async with httpcore.AsyncHTTPProxy(proxy_server) as http:
+    async with httpcore.AsyncHTTPProxy(proxy_server, proxy_mode=proxy_mode) as http:
         method = b"GET"
         url = (b"http", b"example.org", 80, b"/")
         headers = [(b"host", b"example.org")]
         http_version, status_code, reason, headers, stream = await http.request(
             method, url, headers
         )
-        body = await read_body(stream)
+        _ = await read_body(stream)
+
+        assert http_version == b"HTTP/1.1"
+        assert status_code == 200
+        assert reason == b"OK"
+
+
+@pytest.mark.parametrize("proxy_mode", ["TUNNEL_ONLY"])
```

Comment:
Forwarding HTTPS requests doesn't work with mitmproxy or squid, but I'm not sure if it's by design or a problem with our code, any ideas?

Reply:
Do you mean...

- Sending HTTPS Requests through the proxy? Works for me, at least to external hosts - never managed to set up a proxy that could connect to a localhost server with custom certs.
- Connecting to the proxy via HTTPS (ie the proxy URL is https://...)? Never got it to work, mostly because I don’t know of a proxy that supports it/how to set it up on Squid/mitm.

Reply:
I mean the former, sending requests to `https://example.org` via a `mitmproxy` instance fails with a 400 and a mysterious HTML response content reading `HttpException("Invalid request scheme: https")`. Which is why I was guessing it's simply not supported, but I can't find concrete proof of that.

> Works for me, at least to external hosts

Any details on how you're achieving that? I'm basically recreating what the tests are doing with `trustme-cli`. Adding `FORWARD_ONLY` to the parametrization results in exactly the same problem.

Reply:
> Any details on how you're achieving that?

It worked with Squid, haven't tried with MITM proxy though. For mitmproxy, [How mitmproxy works](https://docs.mitmproxy.org/stable/concepts-howmitmproxyworks/) contains very detailed info about what happens when a client issues an HTTPS request to forward, but I can't wrap my ahead around what actually needs to be configured for this to work…

Reply:
Huh, I just tried it with Squid and got:

`b'HTTP/1.1' 501 b'Not Implemented'` and a bunch of headers and HTML from the Squid server.

Here's the script I'm running to test, in case I'm missing something. The Squid setup is as in your comment.

```
import asyncio
import httpcore

async def read_body(stream):
    try:
        body = []
        async for chunk in stream:
            body.append(chunk)
        return b"".join(body)
    finally:
        await stream.aclose()

async def tunnel_proxy(proxy_server, proxy_mode):
    async with httpcore.AsyncHTTPProxy(proxy_server, proxy_mode=proxy_mode) as http:
        method = b"GET"
        url = (b"https", b"example.org", 443, b"/")
        headers = [(b"host", b"example.org")]
        http_version, status_code, reason, headers, stream = await http.request(
            method, url, headers
        )
        print(http_version, status_code, reason, headers)
        body = await read_body(stream)
        print(body)

        assert http_version == b"HTTP/1.1"
        assert status_code == 200
        assert reason == b"OK"


def main():
    proxy_server = (b"http", b"localhost", 3128)
    proxy_mode = "FORWARD_ONLY"
    asyncio.run(tunnel_proxy(proxy_server, proxy_mode))

if __name__ == "__main__":
    main()
```

Reply:
I spent some time researching this and I can't find any definite answer but given what we wrote on the [HTTPX proxy docs](https://www.python-httpx.org/advanced/#http-proxying):

> To make sure that proxies cannot read your traffic, and even if the proxy_url uses HTTPS, it is recommended to use HTTPS and tunnel requests if possible.

The fact that we default to tunnelling HTTPS requests and forward HTTP ones, and that AFAICT neither mitmproxy or Squid implement forwarding HTTPS requests so I'm going to assume it's discouraged but presumably possible to forward HTTPS requests.

I changed the name of the test and made a note to reflect that.

Thanks for looking into it. 👍